dubbo-common/src/main/java/org/apache/dubbo/common/deploy/DeployListenable.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.common.deploy;
+
+import org.apache.dubbo.rpc.model.ScopeModel;
+
+public interface DeployListenable<T extends ScopeModel> {
```

Comment:
Is this necessary? Can we use SPI to inject automatically?

Reply:
The operations of `DeloyerListener` defined in `AbstractDeployer`, changing it will affect `ModuleDeployer`. 
 This problem can fix in the future work about `ModuleLifecycle`, because the definition of `DeployListener` is very similer to `Application/ModuleLifecycle` :

```Java
public interface DeployListener<E extends ScopeModel> {
    /**
     * Useful to inject some configuration like MetricsConfig, RegistryConfig, etc.
     */
    void onInitialize(E scopeModel);

    void onStarting(E scopeModel);

    void onStarted(E scopeModel);

    void onStopping(E scopeModel);

    void onStopped(E scopeModel);

    void onFailure(E scopeModel, Throwable cause);

}
```

```Java
public interface Lifecycle<T>{

    /**
     * If this lifecycle need to initialize.
     */
    boolean needInitialize(T context);

    default void start(T context){}

    default void initialize(T context){}

    default void preDestroy(T context) {}

    default void postDestroy(T context) {}
}
```